File: Trees/Balanced Binary Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 
public class Solution {
     //isn't just balancing the left and right trees. It is performing the same check for every node
    public bool IsBalanced(TreeNode root) {
        if (root == null) return true;
        
        int left = dfs(root.left);
        int right = dfs(root.right);
        int diff = Math.Abs(left - right);
        if (diff > 1) 
        {
            return false;
        }
        if (!IsBalanced(root.left) || !IsBalanced(root.right)) return false;
        
        return true;

    }
    
    public int dfs(TreeNode node){
        if (node == null) return 0;
        return 1 + Math.Max(dfs(node.left),dfs(node.right));
    }
}

File: Trees/Balanced Binary Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
class Height {
  h: number;
  constructor() {
    this.h = 0;
  }
}

function isBalanced(root: TreeNode | null): boolean {
     function checkBalanced(root: TreeNode | null, h: Height): boolean {
         if (!root) 
         {
             h.h = 0;
             return true;
         }

         const lh = new Height();
         const rh = new Height();

         const leftBal = checkBalanced(root.left, lh);
         const rightBal = checkBalanced(root.right, rh);

         h.h = Math.max(lh.h, rh.h) + 1;

         if (leftBal && rightBal && (Math.abs(lh.h - rh.h) <= 1)) return true;

         return false;
     }

     const height = new Height();
     return checkBalanced(root, height);
};
